source/common/common/absl_mutex.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include "common/common/thread_annotations.h"
+
+#include "absl/synchronization/mutex.h"
+
+namespace Envoy {
+namespace Locking {
+
+// Substitute for absl::MutexLock for use in Envoy, with non-const ref
+// style, and more importantly, making it easier to release locks in
+// the middle of scopes. absl::Mutex on its own is risky to use in
+// the context of code that throws exceptions, as you don't want a
+// thrown exception to leave something locked. However
+// absl::MutexLock is hard to use when you need to interleave
+// variable-scopes and lock-scopes, because it has no Unlock().
+//
+// Functionally, you could accomplish this with
+// std::unique_ptr<absl::MutexLock>, but the clang lock analysis
+// does not understand that composition, and would give false locking
+// errors; we need a MutexLock with the functionality needed in the
+// Envoy codebase (slightly beyond absl's) *and* the lock annotations.
+class SCOPED_LOCKABLE MutexLock {
```

Comment:
Seems this is behavior that would be good to have in Abseil. Do you think it makes sense to file an issue in their GH repo to add this?

Reply:
Good idea, will do.

Reply:
RE wrap absl::Mutex in something derived from BasicLockable?  Yes.  Will do in a follow-up I think.  That helps I think when we are talking about a class that might be instantiated in both cross-thread and cross-process contexts; is that right?

RE std::unique_lock?  I don't think so.  The thread annotation library does not seem to find annotations through that barrier, which is why the MutexLock structure needs annotations itself.  In fact MutexLock could easily be built around BasicCopyable or std::mutex for that matter; the key is the annotations on the scoped lock class.

Reply:
> That helps I think when we are talking about a class that might be instantiated in both cross-thread and cross-process contexts; is that right?

Yes, it would be nice if we could figure out a way to have the annotations also work on the cross process locks somehow.